### ISSUE TITLE
remove all special characters from stats outputs

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#2320]
While working on #2311, I've hit headaches because of special characters in stats output.  For example, in kvs stats, we have fields like `#no-op stores`.  So because of that, we have to use jq's special bracket syntax to parse it.

`flux module stats kvs | jq  '.["namespace"]["primary"]["#no-op stores"]'`

No biggie.  But inside sharness, where everything is run under `test_expect_success` (or similar) those single quotes can become a hassle.  I resolved it via just escaping the quotes:

`flux module stats kvs | jq  .[\"namespace\"][\"primary\"][\"#no-op stores\"]`

But then, sometimes we want to do comparison checks.  Normally:

`flux module stats kvs | jq  -e '.["namespace"]["primary"]["#no-op stores"] == 3'`

So now things are just becoming a headache, and sort of unreadable.

If we just have a rule of no special chars in stats output, everything would be easier.

**Comment by chu11** [comment]
oh crap, I just realized, we just added a special char in #2315.

Maybe we should keep `--parse` (see #2311 )

**Comment by chu11** [comment]
oh crap, we just added a special char via #2315 (dash in namespace, which will be in namespace names in stats output).

**Comment by grondo** [comment]
Even dash is a special character?

**Comment by grondo** [comment]
If the quoting in sharness becomes a pain, I usually write a shell function to do the `jq` parsing, and call the shell function from within the quoted sharness script.

Though I do see the irony in complaining about periods in kvs stats keys when there were already keys like `"obj size total (MiB)"`.

**Comment by chu11** [comment]
> Even dash is a special character?

looks like it

```
>flux module stats kvs | jq ".namespace.primary.noop-stores"
jq: error: stores/0 is not defined at <top-level>, line 1:
.namespace.primary.noop-stores                        
jq: 1 compile error

>flux module stats kvs | jq '.["namespace"]["primary"]["noop-stores"]'

```

**Comment by grondo** [comment]
Try quoting just the key that needs it:
```
$ flux module stats kvs | jq '.namespace.primary."#no-op stores"'
0

```

**Comment by grondo** [comment]
Doesn't seem so bad to me that we need to change everything. I don't think it is reasonable to remove every character that `jq` might consider special.

**Comment by chu11** [comment]
> If the quoting in sharness becomes a pain, I usually write a shell function to do the jq parsing, and call the shell function from within the quoted sharness script. This also helps with the "hard to read" problem.

I realized that, but was trying to avoid creating a bunch of random functions.  And when a `flux module stats --parse foo.bar.baz` was being passed into a `flux exec` call, and I just went "ugh ...".

**Comment by grondo** [comment]
Oh, yeah. Good point on that 2nd example...

**Comment by grondo** [comment]
though you can just pipe the output of `flux exec` to `jq` you don't have to actually run `jq` remotely... However, I realize I'm not being too helpful :-) I'm fine with keeping --parse.

**Comment by chu11** [comment]
> though you can just pipe the output of flux exec to jq

True.

But I think the original assumption was that parsing json "by hand" via `--parse` was sort of stupid and `jq` would be better and presumably easy.  But I think `--parse` ended up showing its worth from a lot of these little headaches.

I already closed, #2311, so I'll close this as well.

**Comment by grondo** [comment]
Yeah the stats and `--parse` seem to really be useful for testing so if it makes testing life easier then we shouldn't remove it!
